src/app/Library/CrudPanel/Traits/MorphRelationships.php

```diff
@@ -0,0 +1,210 @@
+     *
+     * @param  array  $field
+     * @return array
+     */
+    private function makeSureMorphSubfieldsAreDefined(array $field)
+    {
+        if (isset($field['relation_type']) && $field['relation_type'] === 'MorphTo') {
+            [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($field['name']);
+            if (! $this->hasFieldWhere('name', $morphTypeFieldName) || ! $this->hasFieldWhere('name', $morphIdFieldName)) {
+                // create the morph fields in the crud panel
+                $field = $this->createMorphToRelationFields($field, $morphTypeFieldName, $morphIdFieldName);
+                foreach ($field['morphOptions'] ?? [] as $morphOption) {
+                    [$key, $label, $options] = $this->getMorphOptionStructured($morphOption);
+                    $field = $this->addMorphOption($field, $key, $label, $options);
+                }
+            }
+        }
+
+        return $field;
+    }
+
+    /**
+     * This function is responsible for setting up the morph fields structure.
+     * Developer can define the morph structure as follows:
+     *  'morphOptions => [
+     *       ['nameOnAMorphMap', 'label', [options]],
+     *       ['App\Models\Model'], // display the name of the model
+     *       ['App\Models\Model', 'label', ['data_source' => backpack_url('smt')]
+     *  ]
+     * OR
+     * ->addMorphOption('App\Models\Model', 'label', ['data_source' => backpack_url('smt')]).
+     *
+     * @param  string|array  $fieldName
+     * @param  string  $key
+     * @param  string|null  $label
+     * @param  array  $options
+     * @return void|array
+     */
+    public function addMorphOption($fieldName, string $key, $label = null, array $options = [])
+    {
+        $morphField = is_array($fieldName) ? $fieldName : $this->fields()[$fieldName];
+
+        $fieldName = $morphField['name'];
+
+        [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($fieldName);
+
+        if (! in_array($morphTypeFieldName, array_column($morphField['subfields'], 'name')) ||
+            ! in_array($morphIdFieldName, array_column($morphField['subfields'], 'name'))) {
+            throw new \Exception('Trying to add morphOptions to a non morph field. Check if field and relation name matches.');
+        }
+
+        [$morphTypeField, $morphIdField] = $morphField['subfields'];
+
+        $morphMap = $morphTypeField['morphMap'];
+
+        if (array_key_exists($key, $morphTypeField['options'] ?? [])) {
+            throw new \Exception('Duplicate entry for «'.$key.'» in addMorphOption().');
+        }
+
+        if (is_a($key, 'Illuminate\Database\Eloquent\Model', true)) {
+            if (in_array($key, $morphMap)) {
+                $key = $morphMap[array_search($key, $morphMap)];
+
+                if (array_key_exists($key, $morphTypeField['options'])) {
+                    throw new \Exception('Duplicate entry for «'.$key.'» in morphOptions');
```

Comment:
This error message is the same 3 times. That should never happen, it defeats the purpose of error messages. If someone comes back to us saying "_i got this error message_" we won't be able to tell exactly where it errored, right? Let's make these error messages more specific, so that they pinpoint what went wrong, or recommend specifically what to fix. The way I see it, each error message should be slightly different, otherwise... it would have been caught by the previous exception, right?